### DISCUSSION TITLE
Custom properties on custom controls

### DISCUSSION THREAD

**Discussion by DirkKramer** [AvaloniaUI/Avalonia#9841] [Q&A] [answered]
I want to create a custom control with custom properties and commands but I cant get it to work.

I am working on a multiplatform (android,windows,etc..) solution in avalonia v11.

I tried the following code:

My user control:  .axaml.cs
 
```
public partial class CustomButton : UserControl
    {

        public static readonly StyledProperty<string> HeaderProperty =
        AvaloniaProperty.Register<CustomButton, string>(nameof(Header), defaultValue: "",
            defaultBindingMode: BindingMode.TwoWay);

        public string Header
        {
            get => GetValue(HeaderProperty);
            set => SetValue(HeaderProperty, value);
        }

        public CustomButton()
        {
            InitializeComponent();
        }

    }
```

My user control: .axaml

```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="RapiDVision.Avalonia.CustomControls.CustomButton">
	<Button >
		<Label
			FontSize="16"
			Content="{Binding Header}"
			HorizontalContentAlignment="Center"	>
		</Label>
	</Button>
</UserControl>

```

My view using the custom control:
```
`
<cc:CustomButton  Header="Settings" ></cc:CustomButton>

`
```

The text 'Settings' wont get displayed. What is the proper way to create custom properties and commands on an custom control.

Thanks in advance!

**Comment by timunie** [comment, marked as answer]
Hi @DirkKramer ,

UserControl is not made for this. Instead, I'd use `TemplatedControl` or `Button` as my base class. See: https://docs.avaloniaui.net/docs/authoring-controls/types-of-control . The benefit of using Button is, you already have all needed properties and events for interaction.

Note: In such a simple case maybe a Style with class `mySpecialButton` would work as well. https://docs.avaloniaui.net/docs/styling/selectors#class

**Reply by DirkKramer** [reply]
The case above is just a simple case to describe the problem, the actual control I want to make is more complex. (a style would not suffice in that case)

Using Button as base class does not resolve the issue. I dont know how to use TemplatedControl, and I cant find many sources for an explanation/example.

**Reply by timunie** [reply]
I can offer this one, but it's not finished yet. https://github.com/timunie/Avalonia.Samples/tree/Controls/RatingControl/src/Avalonia.Samples/CustomControls/RatingControlSample

**Reply by DirkKramer** [reply]
Thanks for that link!!  my control is now working as expected.

**Reply by agillet76** [reply]
@DirkKramer  Can you post your solution to your issue? I am having the same issue with trying to implement a custom control as well.
What did you have to modify to make it work?

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src%2FAvalonia.Samples%2FCustomControls%2FRatingControlSample

**Reply by DirkKramer** [reply]
@agillet76  I dont know if it helps but for my button, this is the button class:
```
 public class FlyoutButton : TemplatedControl
    {
        public static readonly StyledProperty<Geometry> DataProperty =
            AvaloniaProperty.Register<FlyoutButton, Geometry>(
                nameof(Data),
                defaultValue: null);
        public Geometry Data
        {
            get { return GetValue(DataProperty); }
            set { SetValue(DataProperty, value); }
        }


        public static readonly StyledProperty<object> CommandParameterProperty =
         AvaloniaProperty.Register<FlyoutButton, object>(
             nameof(CommandParameter),
             defaultValue: null);
        public object CommandParameter
        {
            get { return GetValue(CommandParameterProperty); }
            set { SetValue(CommandParameterProperty, value); }
        }

        public static readonly StyledProperty<string> HeaderProperty =
            AvaloniaProperty.Register<FlyoutButton, string>(
                nameof(Header),     
                defaultValue: "");
        public string Header
        {
            get { return GetValue(HeaderProperty); }
            set { SetValue(HeaderProperty, value); }
        }

        public static readonly StyledProperty<ICommand> CommandProperty =
         AvaloniaProperty.Register<FlyoutButton, ICommand>(
             nameof(Command));
        public ICommand Command
        {
            get { return GetValue(CommandProperty); }
            set { SetValue(CommandProperty, value); }
        }
    }
```

And this is in the styles.axaml

```
﻿<Styles xmlns="https://github.com/avaloniaui"
		 xmlns:i="clr-namespace:Avalonia.Xaml.Interactivity;assembly=Avalonia.Xaml.Interactivity"
		xmlns:ia="clr-namespace:Avalonia.Xaml.Interactions.Core;assembly=Avalonia.Xaml.Interactions"
		xmlns:controls="clr-namespace:RapiDVision.CustomControls"
		xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
	<Design.PreviewWith>
		<Border Padding="20">
			<controls:FlyoutButton Header="Hello"></controls:FlyoutButton>
		</Border>
	</Design.PreviewWith>

<!--FlyoutButton-->
	<Style Selector="controls|FlyoutButton">
		<Setter Property="Template">
			<ControlTemplate>
				<Button Height="50"
					HorizontalAlignment="Stretch"
					HorizontalContentAlignment="Stretch"
					Background="Transparent"
					Command="{TemplateBinding Command}"
					CommandParameter="{TemplateBinding CommandParameter}">
					<DockPanel HorizontalAlignment="Stretch" VerticalAlignment="Stretch">
						<PathIcon
							Foreground="{DynamicResource IconBrush}"
							Margin="5,0"  Width="20" Height="20" DockPanel.Dock="Left"
							Data="{TemplateBinding Data}" >

							
						</PathIcon>

						<TextBlock Margin="10,0" VerticalAlignment="Center" DockPanel.Dock="Right" Classes="h2" Text="{TemplateBinding Header}"></TextBlock>
					</DockPanel>
				</Button>
			</ControlTemplate>
		</Setter>
	</Style>
</Styles>
```

**Reply by timunie** [reply]
@DirkKramer you can also base on Button, so you only need to implement the Icon stuff I guess. Just fyi.

**Reply by DirkKramer** [reply]
> @DirkKramer you can also base on Button, so you only need to implement the Icon stuff I guess. Just fyi.

For example ` public class FlyoutButton : Button ` ?

**Reply by timunie** [reply]
Exactly 💯

**Reply by agillet76** [reply]
Thanks a lot for the quick reply. I was able to get my custom control working.

**Reply by YannikBartel** [reply]
could you post some axaml of custom control derived from Button? My custom control won't get displayed. I want to set <Button.Content>. Compiles but won't show in previewer or at runtime. @timunie @agillet76

**Reply by timunie** [reply]
Remember to add StyleInclude, see Avalonia.Samples > RatingControl

**Reply by YannikBartel** [reply]
I will look into Styleinclude, thanks. 
But RatingControl is derived from TemplatedControl, not Button directly. Is this same anyway (for style)?

**Reply by timunie** [reply]
It is. Button is a templated control under the hood.
